src/debugAdapter/goDebug.ts

```diff
@@ -141,57 +144,69 @@ function log(msg?: any, ...args) {
 }
 
 class Delve {
+    program: string;
+    remotePath: string;
 	debugProcess: ChildProcess;
 	connection: Promise<RPCConnection>;
 	onstdout: (str: string) => void;
 	onstderr: (str: string) => void;
 
-	constructor(mode: string, program: string, args: string[], cwd: string, env: { [key: string]: string }, buildFlags: string, init: string) {
-		this.connection = new Promise((resolve, reject) => {
-			let serverRunning = false;
-			let dlv = getBinPath('dlv');
-			log('Using dlv at: ', dlv);
-			if (!existsSync(dlv)) {
-				return reject('Cannot find Delve debugger. Ensure it is in your `GOPATH/bin` or `PATH`.');
-			}
-			let dlvEnv: Object = null;
-			if (env) {
-				dlvEnv = {};
-				for (let k in process.env) {
-					dlvEnv[k] = process.env[k];
-				}
-				for (let k in env) {
-					dlvEnv[k] = env[k];
-				}
-			}
-			let dlvArgs = [mode || 'debug'];
-			if (mode === 'exec') {
-				dlvArgs = dlvArgs.concat([program]);
-			}
-			dlvArgs = dlvArgs.concat(['--headless=true', '--listen=127.0.0.1:2345', '--log']);
-			if (buildFlags) {
-				dlvArgs = dlvArgs.concat(['--build-flags=' + buildFlags]);
-			}
-			if (init) {
-				dlvArgs = dlvArgs.concat(['--init=' + init]);
-			}
-			if (args) {
-				dlvArgs = dlvArgs.concat(['--', ...args]);
-			}
-
-			let dlvCwd = dirname(program);
-			try {
-				if (lstatSync(program).isDirectory()) {
-					dlvCwd = program;
-				}
-			} catch (e) { }
-			this.debugProcess = spawn(dlv, dlvArgs, {
-				cwd: dlvCwd,
-				env: dlvEnv,
-			});
-
-			function connectClient() {
-				let client = Client.$create(2345, '127.0.0.1');
+	constructor(mode: string, remotePath: string, port: number, host:string, program: string, args: string[], cwd: string, env: { [key: string]: string }, buildFlags: string, init: string) {
+        this.program = program;
+        this.remotePath = remotePath;
+        this.connection = new Promise((resolve, reject) => {
+            let serverRunning = false;
+            if (mode === 'debug_remote') {
+                this.debugProcess = null;
+                serverRunning = true;  //assume server is running when in debug_remote mode
+                connectClient(port, host);
+            }
+            else 
```

Comment:
To avoid the additional nesting here - can the remote path complete with a return and then keep the local path at top level of the Promise?  The `if (this.debugProcess != null)` further down can then be removed as well I believe.